gobblin-data-management/src/main/java/org/apache/gobblin/data/management/source/LoopingDatasetFinderSource.java

```diff
@@ -79,59 +88,122 @@ public LoopingDatasetFinderSource(boolean drilldownIntoPartitions) {
     return Lists.newArrayList(getWorkunitStream(state).getMaterializedWorkUnitCollection());
   }
 
+  /**
+   * Collect all failed workunit in previous run into a set.
+   *
+   * Workunits that exceed max retry limit will not be added into failedPreviousWorkUnits
+   * and rollovered into next quota-available execution.
+   *
+   * @param retriedUrns Track all dataset/partition's urns that is included in the retried workunits, so that in the
+   *                    case where the number of remaining dataset/partitions are less than the maximum number of
+   *                    workunits can be processed by single execution, a retried workunit won't be processed twice
+   */
+  private List<WorkUnit> getPreviousFailedWorkUnits(List<WorkUnitState> previousWorkUnitStates, int maxRetries,
+      Set<String> retriedUrns) {
```

Comment:
There is no need to pass in a `Set` to this method that gets modified. The `Set` can be populated outside this method (better yet, in a separate method) from the `List<WorkUnit>` returned by this method.

Generally, it's desirable for methods not to have side effects. It's best if a method takes inputs and returns an output based on the inputs without modifying any of the inputs.

Reply:
Is side effect in java always considered as bad practice?

Reply:
Almost all rules have exceptions. I just think as a general rule of thumb, you should avoid writing methods with side effects. This is the main idea behind functional programming, where outputs of a function only depend on inputs and functions have no side effects.